packages/react-dom-bindings/src/server/ReactDOMServerFormatConfig.js

```diff
@@ -2233,57 +2256,105 @@ function escapeJSObjectForInstructionScripts(input: Object): string {
   });
 }
 
+const precedencePlaceholderStart = stringToPrecomputedChunk(
+  '<style data-rprec="',
+);
+const precedencePlaceholderEnd = stringToPrecomputedChunk('"></style>');
+
 export function writeInitialResources(
   destination: Destination,
   resources: Resources,
   responseState: ResponseState,
 ): boolean {
-  const explicitPreloadsTarget = [];
-  const remainingTarget = [];
+  const target = [];
 
-  const {precedences, explicitPreloads, implicitPreloads} = resources;
+  const {
+    fontPreloads,
+    precedences,
+    usedStylePreloads,
+    scripts,
+    usedScriptPreloads,
+    explicitStylePreloads,
+    explicitScriptPreloads,
+  } = resources;
+
+  fontPreloads.forEach(r => {
+    // font preload Resources should not already be flushed so we elide this check
+    pushLinkImpl(target, r.props, responseState);
+    r.flushed = true;
+  });
+  fontPreloads.clear();
 
   // Flush stylesheets first by earliest precedence
-  precedences.forEach(precedenceResources => {
-    precedenceResources.forEach(resource => {
-      // resources should not already be flushed so we elide this check
-      pushLinkImpl(remainingTarget, resource.props, responseState);
-      resource.flushed = true;
-      resource.inShell = true;
-      resource.hint.flushed = true;
-    });
+  precedences.forEach((p, precedence) => {
+    if (p.size) {
+      p.forEach(r => {
+        // resources should not already be flushed so we elide this check
+        pushLinkImpl(target, r.props, responseState);
+        r.flushed = true;
+        r.inShell = true;
+        r.hint.flushed = true;
+      });
+      p.clear();
+    } else {
+      target.push(
+        precedencePlaceholderStart,
+        escapeTextForBrowser(stringToChunk(precedence)),
+        precedencePlaceholderEnd,
+      );
+    }
   });
 
-  explicitPreloads.forEach(resource => {
-    if (!resource.flushed) {
-      pushLinkImpl(explicitPreloadsTarget, resource.props, responseState);
-      resource.flushed = true;
+  usedStylePreloads.forEach(r => {
+    // style preloads could very likely have been flushed already so we check
+    if (!r.flushed) {
+      pushLinkImpl(target, r.props, responseState);
+      r.flushed = true;
+    }
+  });
+  usedStylePreloads.clear();
+
+  scripts.forEach(r => {
```

Comment:
The preamble notes didn't have anything about scripts in it.

I think we said that we don't want to emit scripts in the preamble, except the external fizz runtime. Because they can end up flooding the network ahead of the styles that are going to come soon.

Reply:
They're not needed for display. The exception is the fizz runtime which will be needed to display boundaries.

Reply:
Seems like we should prep the code for it though. Otherwise, what's the difference between Immediate and Initial?

Reply:
we don't flush styles in immediate